File: Coursera/Algorithms_Part_2/Ex13EdgeWeightedGraph.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.Bag;
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdOut;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class Ex13EdgeWeightedGraph {

    private final int V;
    private int E;
    private Bag<Ex12Edge>[] adj;

    public Ex13EdgeWeightedGraph(int V) {
        this.V = V;
        this.E = 0;
        this.adj = (Bag<Ex12Edge>[]) new Bag[V];
        for (int i = 0; i < V; i++)
            adj[i] = new Bag<>();
    }

    public Ex13EdgeWeightedGraph(In in) {
        this.V = in.readInt();
        this.E = in.readInt();
        this.adj = (Bag<Ex12Edge>[]) new Bag[V];
        for (int i = 0; i < V; i++)
            adj[i] = new Bag<>();

        while (!in.isEmpty()) {
            addEdge(new Ex12Edge(in.readInt(), in.readInt(), in.readDouble()));
            this.E--;
        }
    }

    public void addEdge(Ex12Edge e) {
        int v = e.either();
        int w = e.other(v);
        adj[v].add(e);
        adj[w].add(e);
        this.E++;
    }

    public Iterable<Ex12Edge> adj(int v) {
        return adj[v];
    }

    public Iterable<Ex12Edge> edges() {
        return () -> new Iterator<Ex12Edge>() {
            private int current = 0;

            @Override
            public boolean hasNext() {
                return current < V();
            }

            @Override
            public Ex12Edge next() {
                if (!hasNext()) throw new NoSuchElementException();
                if (!adj[current].iterator().hasNext()) current++;
                return adj[current].iterator().next();
            }
        };
    }

    public int V() {
        return V;
    }

    public int E() {
        return E;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append(V).append(" vertices ");
        builder.append(E).append(" edges");
        builder.append("\n");
        for (int i = 0; i < V(); i++) {
            builder.append(i).append(":");
            for (Ex12Edge e : adj(i))
                builder.append(e).append(" ");
            builder.append("\n");
        }
        /* for (Ex12Edge edge : edges())
            builder.append(edge.toString()).append("\n");*/
        return builder.toString();
    }

    public static void main(String[] args) {
        In in = new In("tinyEWG.txt");
        Ex13EdgeWeightedGraph graph = new Ex13EdgeWeightedGraph(in);
        StdOut.println(graph.toString());
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex13EdgeWeightedGraph.py
from Ex12Edge import Edge


class EdgeWeightedGraph:

    def __init__(self, v):
        self.v = v
        self.e = 0
        self.adj = [[] for _ in range(self.v)]

    def init_from(self, file_name):

        with open(file_name) as f:

            self.v = int(f.readline())
            self.e = int(f.readline())

            self.adj = [[] for _ in range(self.v)]

            for _ in range(self.e):
                v, w, weight = f.readline().split()
                self.add_edge(int(v), int(w), float(weight))

    # noinspection PyTypeChecker
    def add_edge(self, v, w, weight):
        edge = Edge(v, w, weight)
        self.adj[v].append(edge)
        self.adj[w].append(edge)

    def get_adj(self, v):
        for edge in self.adj[v]:
            yield edge

    def get_v(self):
        return self.v

    def get_e(self):
        return self.e

    def __str__(self):
        builder = f'Vertices: {self.v} Edges: {self.e}\n'
        for i in range(self.get_v()):
            builder += f'{i}:'
            for edge in self.get_adj(i):
                builder += f' {edge.other(i)},{edge.get_weight()} '
            builder += '\n'
        return builder


def test_edge_weighted_graph():
    graph = EdgeWeightedGraph(0)
    graph.init_from('tinyEWG.txt')
    print(graph)


if __name__ == '__main__':
    test_edge_weighted_graph()
